src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -3834,38 +3846,39 @@ void EmitStackPop(params string[] args)
             static string FormatN(string format, int count) =>
                 string.Concat(from i in Enumerable.Range(0, count)
                               select string.Format(format, i));
+        }
 
-            /// <summary>Emits a timeout check.</summary>
-            void EmitTimeoutCheckIfNeeded()
+        /// <summary>Emits a timeout check if the regex timeout wasn't explicitly set to infinite.</summary>
+        /// <remarks>
+        /// Regex timeouts exist to avoid catastrophic backtracking.  The goal with timeouts isn't to be accurate to the timeout value,
+        /// but to ensure that significant backtracking can be stopped.  As such, we allow for up to O(n) work in the length of the input
+        /// between checks, which means we emit checks anywhere backtracking is introduced, such that every check can have O(n) work
+        /// associated with it.  This means checks:
+        /// - when restarting the whole match evaluation at a new index
```

Comment:
These remarks are great! I saw that you are also updating the docs on Regex.Timeout.cs but I really like how in here we very specifically call out exactly when do we perform timeout checks today, while in RegexTimeout.cs docs you have a more general guarantees. Do you think it might be interesting from a consumer's perspective to know exactly when to expect timeout checks to happen? Or are the comments with the guarantees sufficient?

Reply:
These remarks are focused more on implementation; I don't think we need to guarantee this level of granularity, just that we reserve the right to do up to O(n) work between timeout checks.